packages/open-truss/src/configuration/engine-v1.tsx

```diff
@@ -27,21 +31,29 @@ export function engineV1(
   config: WorkflowV1,
   data: YamlType,
 ): RenderingEngine {
-  const renderFrames = (frames: FrameV1[]): ReactTree => {
-    return frames.map(({ view, data, frames: subFrame }, i) => {
-      const { component, props } = view
-      const Component = COMPONENTS[component]
-      return subFrame === undefined ? (
-        <Component key={i} data={data} config={config} {...props} />
-      ) : (
-        <Component key={i} data={data} config={config} {...props}>
-          {renderFrames(subFrame)}
-        </Component>
-      )
-    })
+  const renderFrames = async (frames: FrameV1[]): Promise<ReactTree> => {
+    return (() => {
+      return frames.map(async ({ view, data, frames: subFrame }, i) => {
```

Comment:
I think you’ll want to wrap this in a Promise.all maybe? Typically keeps things in order when you map over an array of promises.

Reply:
I agree. I tried doing this and the types got upset 🤔

```tsx
  const renderFrames = (frames: FrameV1[]): Promise<ReactTree> => {
    const foo = Promise.all(frames.map(async ({ view, data, frames: subFrame }, i) => {
      const { component, props: viewProps } = view
      const Component = COMPONENTS[component]
      const props = {
        key: i,
        data,
        config,
        ...viewProps,
      }

      if (subFrame === undefined) {
        return await Component({ ...props })
      } else {
        const children = <>{renderFrames(subFrame)}</>
        return await Component({ ...props, children })
      }
    }))
    return foo
  }
```

> Type 'Promise<Element[]>' is not assignable to type 'Promise<ReactTree>'.

Reply:
Let me play around with this to see if I can get this to work.

Reply:
I think I got this working here https://github.com/open-truss/open-truss/pull/60/commits/2103216c2bc12ce7b764864c5605183906831c51